File: cr_ultimate_gantt_pro/static/src/js/find_div.js
const fs = require('fs');
const content = fs.readFileSync('c:/Users/charm/Documents/odoo/odoo-18.0/custom/addons-data/addons/cr_ultimate_gantt_pro/static/src/js/ultimate_gantt_bundle.js', 'utf8');
const lines = content.split('\n');

const startIndex = 735; // 0-indexed (line 736)
const endIndex = 1332; // 0-indexed (line 1333)

const stack = [];

for (let i = startIndex; i <= endIndex; i++) {
    const line = lines[i];
    // Very simple matching for <div> and </div>
    // This isn't perfect but helps visualize
    let match;
    const openRegex = /<div(?=[\s>])/gi;
    while ((match = openRegex.exec(line)) !== null) {
        // Exclude self-closing <div ... />
        const closeIndex = line.indexOf('>', match.index);
        const selfCloseIndex = line.indexOf('/>', match.index);
        if (selfCloseIndex !== -1 && selfCloseIndex === closeIndex - 1) {
            // self-closing
            continue;
        }
        stack.push(`line ${i+1}`);
    }
    
    const closeRegex = /<\/div>/gi;
    while ((match = closeRegex.exec(line)) !== null) {
        if (stack.length > 0) {
            stack.pop();
        } else {
            console.log(`Unmatched </div> at line ${i+1}`);
        }
    }
}

console.log(`Remaining unclosed <div> count: ${stack.length}`);
if (stack.length > 0) {
    console.log(`They were opened at:`);
    for (let s of stack) {
        console.log(s);
    }
}

File: cr_ultimate_gantt_pro/static/src/js/find_div.py
from html.parser import HTMLParser

class Tracker(HTMLParser):
    def __init__(self):
        super().__init__()
        self.stack = []
        self.output = []

    def handle_starttag(self, tag, attrs):
        self.stack.append((tag, self.getpos()[0]))

    def handle_endtag(self, tag):
        if self.stack and self.stack[-1][0] == tag:
            self.stack.pop()
        else:
            self.output.append(f"Mismatch at {self.getpos()[0]}: Expected </{self.stack[-1][0]}> (opened at {self.stack[-1][1]}), got </{tag}>")

parser = Tracker()
with open('c:/Users/charm/Documents/odoo/odoo-18.0/custom/addons-data/addons/cr_ultimate_gantt_pro/static/src/js/ultimate_gantt_bundle.js', 'r', encoding='utf-8') as f:
    content = f.read()

start = content.find("static template = xml`") + len("static template = xml`")
end = content.find("`;\n    static components = { UltimateGanttRenderer,")
if end == -1:
    end = content.find("`;", start)

xml_str = content[start:end].strip()

parser.feed(xml_str)
for o in parser.output[:5]:
    print(o)
if not parser.output:
    print("Parsed successfully!")
